File: game-of-life/kotlin-arrow-game-of-life/src/main/kotlin/GamePlay.kt
import arrow.core.None
import arrow.core.Option
import arrow.core.Some
import arrow.fx.IO

object GamePlay {
    fun calculateNextGeneration(matrix: Matrix): IO<Matrix> = IO.just(nextGeneration(matrix))

    private fun nextGeneration(currentGeneration: Matrix): Matrix =
            currentGeneration.mapIndexed { rowPosition, row ->
                row.mapIndexed { columnPosition, cell ->
                    getNextGenerationCellStatusFor(cell, rowPosition, columnPosition, currentGeneration)
                }
            }

    private fun getNextGenerationCellStatusFor(cellStatus: CellStatus, row: Int, column: Int, matrix: Matrix): CellStatus {
        val numberOfNeighboursFor = getNumberOfNeighboursFor(row, column, matrix)
        return when {
            (numberOfNeighboursFor == 2 || numberOfNeighboursFor == 3) && cellStatus == Alive -> Alive
            (numberOfNeighboursFor == 3) && cellStatus == Dead -> Alive
            else -> Dead
        }
    }

    private fun getNumberOfNeighboursFor(row: Int, column: Int, matrix: Matrix): Int =
            getNeighboursValueUsing(row - 1, column - 1, matrix) +
                    getNeighboursValueUsing(row - 1, column, matrix) +
                    getNeighboursValueUsing(row - 1, column + 1, matrix) +
                    getNeighboursValueUsing(row, column - 1, matrix) +
                    getNeighboursValueUsing(row, column + 1, matrix) +
                    getNeighboursValueUsing(row + 1, column - 1, matrix) +
                    getNeighboursValueUsing(row + 1, column, matrix) +
                    getNeighboursValueUsing(row + 1, column + 1, matrix)

    private fun getNeighboursValueUsing(row: Int, column: Int, matrix: Matrix): Int =
            getCellFrom(matrix, row, column).fold(
                    { 0 },
                    { cell ->
                        when (cell) {
                            Alive -> 1
                            Dead -> 0
                        }
                    })

    private fun getCellFrom(matrix: Matrix, rowPosition: Int, columnPosition: Int): Option<CellStatus> {
        matrix.getOrNull(rowPosition)?.let { row ->
            row.getOrNull(columnPosition)?.let { cell ->
                return Some(cell)
            }
        }
        return None
    }
}

File: game-of-life/swift-bow-game-of-life/swift-bow-game-of-life/GamePlay.swift
//
//  GamePlay.swift
//  swift-bow-game-of-life
//
//  Created by Fabrizio Duroni on 30.09.20.
//

import Foundation
import Bow

func nextGeneration(currentGeneration: Matrix) -> Matrix {
    return currentGeneration.enumerated().map { (rowPosition, row) in
        row.enumerated().map { (columnPosition, cell) in
            getNextGenerationCellStatusFor(cellStatus: cell, row: rowPosition, column: columnPosition, matrix: currentGeneration)
        }
    }
}

private func getNextGenerationCellStatusFor(cellStatus: CellStatus, row: Int, column: Int, matrix: Matrix) -> CellStatus {
    let numberOfNeighboursFor = getNumberOfNeighboursFor(row: row, column: column, matrix: matrix)
    
    if ((numberOfNeighboursFor == 2 || numberOfNeighboursFor == 3) && cellStatus == .Alive) {
        return .Alive
    }
    
    if ((numberOfNeighboursFor == 3) && cellStatus == .Dead) {
        return .Alive
    }
    
    return .Dead
}

private func getNumberOfNeighboursFor(row: Int, column: Int, matrix: Matrix) -> Int {
    return getNeighboursValueUsing(row: row - 1, column: column - 1, matrix: matrix) +
        getNeighboursValueUsing(row: row - 1, column: column, matrix: matrix) +
        getNeighboursValueUsing(row: row - 1, column: column + 1, matrix: matrix) +
        getNeighboursValueUsing(row: row, column: column - 1, matrix: matrix) +
        getNeighboursValueUsing(row: row, column: column + 1, matrix: matrix) +
        getNeighboursValueUsing(row: row + 1, column: column - 1, matrix: matrix) +
        getNeighboursValueUsing(row: row + 1, column: column, matrix: matrix) +
        getNeighboursValueUsing(row: row + 1, column: column + 1, matrix: matrix)
}

private func getNeighboursValueUsing(row: Int, column: Int, matrix: Matrix) -> Int {
    getCellFrom(matrix: matrix, rowPosition: row, columnPosition: column).fold(
        { 0 },
        { (cell: CellStatus) in
            switch(cell) {
            case .Alive: return 1
            case .Dead: return 0
            }
        }
    )
}

private func getCellFrom(matrix: Matrix, rowPosition: Int, columnPosition: Int) -> Option<CellStatus> {
    if matrix.indices.contains(rowPosition) && matrix[rowPosition].indices.contains(columnPosition) {
        return Option.some(matrix[rowPosition][columnPosition])
    }
    
    return Option.none()
}
